File: api/db/seed.py
import datetime
import os

from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.engine import Connection, create_engine

from api.db.db import parse_cdb_uri
from api.db.models import MachineStatus, MachineType, metadata_obj
from api.machine import Machine

from .models import machine

FLOORS = [5, 8, 11, 14, 17]


def seed_machines(c: Connection):
    """Seeds the database with a set of machines, hard-coded."""
    ms = []
    for f in FLOORS:
        ms.extend(
            [
                Machine(
                    id=f"{f}:0",
                    floor=f,
                    pos=0,
                    duration=datetime.timedelta(minutes=30),
                    type=MachineType.washer,
                    status=MachineStatus.idle,
                ).dict(),
                Machine(
                    id=f"{f}:1",
                    floor=f,
                    pos=1,
                    duration=datetime.timedelta(minutes=30),
                    type=MachineType.washer,
                    status=MachineStatus.idle,
                ).dict(),
                Machine(
                    id=f"{f}:2",
                    floor=f,
                    pos=2,
                    duration=datetime.timedelta(minutes=40),
                    type=MachineType.dryer,
                    status=MachineStatus.idle,
                ).dict(),
                Machine(
                    id=f"{f}:3",
                    floor=f,
                    pos=3,
                    duration=datetime.timedelta(minutes=40),
                    type=MachineType.dryer,
                    status=MachineStatus.idle,
                ).dict(),
            ]
        )

    ins = insert(machine).values(ms).on_conflict_do_nothing()
    c.execute(ins)


if __name__ == "__main__":
    from dotenv import load_dotenv

    load_dotenv()
    engine = create_engine(parse_cdb_uri(os.environ.get("DB_URI")), echo=True)
    metadata_obj.create_all(engine)
    with engine.connect() as db:
        seed_machines(db)

File: api/app/redis/seed.lua
-- Lua script to seed machines in Redis

local WASHER_DURATION = 1800
local DRYER_DURATION = 2400
local FLOORS = { 5, 8, 11, 14, 17 }

for _, floor in ipairs(FLOORS) do
	-- seed the washers
	for pos = 0, 1 do
		redis.call("JSON.SET", "machine"..":"..floor..":"..pos, ".", string.format(
			[[
			{
				"floor": %i,
				"pos": %i,
				"type": "washer",
				"status": "idle",
				"duration": %i,
				"last_started_at": "1970-01-01T00:00:00Z"
			}
			]], floor, pos, WASHER_DURATION
		))
	end
	-- and the dryers
	for pos = 2, 3 do
		redis.call("JSON.SET", "machine"..":"..floor..":"..pos, ".", string.format(
			[[
			{
				"floor": %i,
				"pos": %i,
				"type": "dryer",
				"status": "idle",
				"duration": %i,
				"last_started_at": "1970-01-01T00:00:00Z"
			}
			]], floor, pos, DRYER_DURATION
		))
	end
end
